keps/sig-network/4770-endpointslice-controller-flexibility/README.md

```diff
@@ -0,0 +1,499 @@
+// It also return which address types are handled.
+func DesiredEndpointSlicesFromEndpoints(
+	endpoints *corev1.Endpoints,
+	maxEndpointsPerSubset int32,
+) ([]*EndpointPortAddressType, sets.Set[discovery.AddressType])
+
+type LabelsAnnotationsFromEndpoints struct {
+	Endpoints *corev1.Endpoints
+}
+
+func (lafe *LabelsAnnotationsFromEndpoints) SetLabelsAnnotations(logger klog.Logger, epSlice *discovery.EndpointSlice, controllerName string) (map[string]string, map[string]string, bool) 
+```
+
+Example of usage replacing the current `Reconcile` call from the EndpointSlice Mirroring Controller ([pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go#L334](https://github.com/kubernetes/kubernetes/blob/v1.30.2/pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go#L334)):
+```golang
+desiredEndpoints, supportedAddressesTypes := endpointslicerec.DesiredEndpointSlicesFromEndpoints(endpoints, r.maxEndpointsPerSubset)
+return rec.Reconcile(
+    logger,
+    endpoints,
+    endpoints,
+    desiredEndpoints,
+    existingSlices,
+    supportedAddressesTypes,
+    nil, // EndpointSlice Mirroring doesn't use traffic distribution.
+    labelsAnnotationsFromEndpoints.SetLabelsAnnotations,
+    time.Time{}, // So endpoints.kubernetes.io/last-change-trigger-time will be removed.
+)
+```
+
+#### Metrics
+
+Most of the metrics exposed by the EndpointSlice and EndpointSlice Mirroring Controllers are similar. The EndpointSlice Controller exposes `endpointslices_changed_per_sync`, `syncs` and `services_count_by_traffic_distribution` that are not exposed by the EndpointSlice Mirroring Controller. The EndpointSlice Mirroring Controller exposes `endpoints_updated_per_sync`, `addresses_skipped_per_sync` and `endpoints_sync_duration` that are not exposed by the EndpointSlice Controller. Metrics exposed will need to be merged together, and the metrics package should be changed to support any type of resources (pods/service, endpoints...).
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+* Tests in `pkg/controller/endpointslicemirroring/reconciler_test.go` will be moved to `staging/src/k8s.io/endpointslice`
+
+##### Integration tests
+
+- Usage of `service.kubernetes.io/endpoint-controller-name` on services
+    * A service is created with the label and the service has then no endpoints neither endpointslices. Then service is updated removing the label and the service has now endpoints and endpointslices.
+    * A service is created without the label, the service has endpoints and endpointslices. Then service is updated with the label and the service has no longer any endpoints nor endpointslices.
+
+##### e2e tests
+
+TDB
```

Comment:
Sorry, my previous comment was confusing, we need at least one or two e2e test that implementation use to validate the expected behavior, implementations must be consistent

Reply:
this is still missing, 


"A service is created with the label and an external controller creates EndpointSlices for that Service, Pods connecting to that service should be able to connect to the custom crafted endpoint"

For this we just can create the fake endpointslices manually on the e2e test emulating the behavior of an external controller and pointing them to some random pod, there are already some tests with this logic you can reuse

Reply:
Thank you for the suggestion, I added it.